### ISSUE TITLE
error: Could not compile `hex-literal`

### ISSUE THREAD

**Issue by anagutjor** [paritytech/polkadot#335]
Hello. I am trying to install polkadot in a fresh install of Ubuntu 18.04. I am strictly following the official [wiki](https://wiki.polkadot.network/en/latest/polkadot/node/guides/how-to-validate/). I have not used any other command outside from this guide other than first creating my sudo user and doing `sudo apt update`, `sudo apt upgrade` and `sudo apt install git build-essential libssl-dev`.

However, when I do `./scripts/build.sh` I encounter the following error (I decided to post the complete logs)

```
*** Building wasm binaries in runtime/wasm
    Updating git repository `https://github.com/paritytech/integer-sqrt-rs.git`
    Updating git repository `https://github.com/paritytech/substrate`
  Downloaded byteorder v1.3.1
  Downloaded integer-sqrt v0.1.2
  Downloaded bitvec v0.8.0
  Downloaded num-traits v0.2.6
  Downloaded parity-codec v3.4.0
  Downloaded hash-db v0.12.2
  Downloaded quote v0.6.12
  Downloaded paste v0.1.5
  Downloaded parity-codec-derive v3.3.0
  Downloaded serde v1.0.90
  Downloaded blake2-rfc v0.2.18
  Downloaded primitive-types v0.2.1
  Downloaded rustc_version v0.2.3
  Downloaded bitmask v0.5.0
  Downloaded proc-macro2 v0.4.27
  Downloaded hex-literal v0.1.4
  Downloaded proc-macro-crate v0.1.3
  Downloaded safe-mix v1.0.0
  Downloaded arrayvec v0.4.10
  Downloaded semver v0.9.0
  Downloaded syn v0.15.33
  Downloaded unicode-xid v0.1.0
  Downloaded impl-codec v0.2.0
  Downloaded proc-macro-hack v0.4.1
  Downloaded hash256-std-hasher v0.12.2
  Downloaded uint v0.6.1
  Downloaded constant_time_eq v0.1.3
  Downloaded rustc-hex v2.0.1
  Downloaded hex-literal-impl v0.1.2
  Downloaded toml v0.4.10
  Downloaded proc-macro-hack v0.5.4
  Downloaded paste-impl v0.1.5
  Downloaded nodrop v0.1.13
  Downloaded semver-parser v0.7.0
  Downloaded fixed-hash v0.3.0
  Downloaded static_assertions v0.2.5
  Downloaded proc-macro-hack-impl v0.4.1
  Downloaded crunchy v0.2.1
   Compiling serde v1.0.90
   Compiling proc-macro2 v0.4.27
   Compiling unicode-xid v0.1.0
   Compiling syn v0.15.33
   Compiling arrayvec v0.4.10
   Compiling semver-parser v0.7.0
   Compiling crunchy v0.2.1
   Compiling byteorder v1.3.1
   Compiling nodrop v0.1.13
   Compiling rustc-hex v2.0.1
   Compiling static_assertions v0.2.5
   Compiling hash-db v0.12.2
   Compiling num-traits v0.2.6
   Compiling constant_time_eq v0.1.3
   Compiling proc-macro-hack-impl v0.4.1
   Compiling integer-sqrt v0.1.2
   Compiling bitmask v0.5.0
   Compiling integer-sqrt v0.1.0 (https://github.com/paritytech/integer-sqrt-rs.git#886e9cb9)
   Compiling bitvec v0.8.0
   Compiling semver v0.9.0
   Compiling fixed-hash v0.3.0
   Compiling proc-macro-hack v0.4.1
   Compiling rustc_version v0.2.3
   Compiling hex-literal-impl v0.1.2
   Compiling toml v0.4.10
   Compiling quote v0.6.12
   Compiling blake2-rfc v0.2.18
   Compiling sr-std v1.0.0 (https://github.com/paritytech/substrate?branch=v1.0#0d4e5a10)
   Compiling sr-io v1.0.0 (https://github.com/paritytech/substrate?branch=v1.0#0d4e5a10)
   Compiling safe-mix v1.0.0
   Compiling hash256-std-hasher v0.12.2
   Compiling uint v0.6.1
   Compiling hex-literal v0.1.4
error: proc-macro derive panicked
  --> /home/ana/.cargo/registry/src/github.com-1ecc6299db9ec823/hex-literal-0.1.4/src/lib.rs:38:1
   |
38 | / proc_macro_expr_decl! {
39 | |     /// Macro for converting hex string to byte array at compile time
40 | |     hex! => hex_impl
41 | | }
   | |_^
   |
   = help: message: assertion failed: `(left == right)`
             left: `Some("#[allow(unused,")`,
            right: `Some("#[allow(unused")`
   = note: this warning originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error

error: Could not compile `hex-literal`.

To learn more, run the command again with --verbose.
```
I have read on this other [issue](https://github.com/paritytech/substrate/issues/3134) on GitHub that I should do `rustup default nightly-2019-07-14`. But then, another error is thrown up with the same command, `./scripts/build.sh`: 
```
*** Building wasm binaries in runtime/wasm
   Compiling serde v1.0.90
   Compiling proc-macro2 v0.4.27
   Compiling unicode-xid v0.1.0
   Compiling syn v0.15.33
   Compiling arrayvec v0.4.10
   Compiling semver-parser v0.7.0
   Compiling crunchy v0.2.1
   Compiling nodrop v0.1.13
error[E0463]: can't find crate for `core`
  |
  = note: the `wasm32-unknown-unknown` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `nodrop`.

To learn more, run the command again with --verbose.
```
I don't know how to fix this last error. If I do `rustup target add wasm32-unknown-unknown --toolchain nightly` it returns:
```
info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
```

Do you have any ideas? Any help appreciated.

**Comment by Kianenigma** [comment]
please see https://github.com/paritytech/substrate/issues/3134. 

Please re-open if your issue is not solved.

**Comment by debuggor** [comment]
I met the same problem, but still didn't solve it.

**Comment by soc1c** [comment]
the hack for older polkadot branches is to remove the wasm runtime cargo-lock which forces bumping hex-literal.

```
rm ./runtime/wasm/Cargo.lock
./scripts/build.sh
```

This could have other side-effects, works for me though.

**Comment by soc1c** [comment]
compare https://github.com/substrate-developer-hub/substrate-package/pull/10

**Comment by vindberg** [comment]
Anyone have more insight into this? Is it okay to use the "hack"? When will this be resolved completely?

Thanks.

**Comment by kianenigma** [comment]
1. The reason that I closed this was that this is resolved in the latest master and only contains the v0.4 branch. The root of the issue is an outdated hex-literal dependency which cannot be compiled with nightly toolchains after a certain date. The latest master already updated this dependency and compliers perfectly fine with both toolchains. As a matter of fact, `build.sh` is also gone in master and wasm binaries are built as a part of the main build command.

2. As for the old branch, @soc1c's suggestion works fine and it is okay to use it.I just tried it out as well and:
  - (maybe not needed) manually bump the hex-literal versions to `0.2` in `Cargo.toml`
  - Remove `Cargo.lock` (`/` and `runtime/wasm`)

`./build.sh` should work fine now.
